src/ray/core_worker/reference_count.h

```diff
@@ -351,8 +363,13 @@ class ReferenceCounter {
     /// - Pending submitted tasks that depend on the object.
     /// - ObjectIDs that we own, that contain this ObjectID, and that are still
     ///   in scope.
+    /// - ObjectIDs that we are borrowing, that contain this ObjectID, and that
+    ///   are still in scope.
     size_t RefCount() const {
-      return local_ref_count + submitted_task_ref_count + contained_in_owned.size();
+      return local_ref_count + submitted_task_ref_count + contained_in_owned.size() +
+                     contained_in_borrowed_id.has_value()
```

Comment:
Why switch to a 0/1 here?

Reply:
This is to deal with nested object IDs. Before, if object A contained object B, then the owner of A would always find out about A before B. But with the new protocol, the borrower of B can bypass the reference tree, so the owner can find out about B first. So we shouldn't consider B's ref count to be 0 until the owner also finds out about A.

To be honest, I'm not completely sure why this didn't come up before. If you think it's too complicated, we could scrap the changes to the ref counting protocol and just have the borrower's raylet periodically check if the owner still has the object in scope instead.